### PR TITLE
gh-118950: Fix SSLProtocol.connection_lost not being called when OSError is thrown on asyncio.write.

### PR DESCRIPTION
**Pull request by cjavad**
Let _SSLProtocolTransport.is_closing reflect SSLProtocol internal transport closing state so that StreamWriter.drain will invoke sleep(0) which calls connection_lost and correctly notifies waiters of connection lost. (#118950)

In an existing comment in [`asyncio/streams.py`](https://github.com/python/cpython/blob/5b941e57c71d7d0ab983d81a169f892662cfe446/Lib/asyncio/streams.py#L385) there is some logic that correctly ensures that the `connection_lost` function is always called for users that keep writing and draining in some other context. 

The issue here is that this code only runs when the transport is marked as closing, which in the case of the _SSLTransportProtocol only happened after connection_lost has been called.

My proposed fix is for _SSLTransportProtocol.is_closing to also return true when its inner transport is marked as closing, when connection_lost is called the inner transport is removed and instead the _closed flag is set.

No additional tests have been added (yet). 

* Issue: gh-118950

### CODE REVIEW

**Comment by gvanrossum** [comment]
@kumaraditya303 Do you have time to review this? I know nothing about SSL. Who else could we ask?

**Comment by kumaraditya303** [comment]
Please add some tests, I'll try to find time to review in following week.

**Comment by Dreamsorcerer** [comment]
I'm not familiar with the SSL code, but the change looks reasonable to me and certainly seems to fix the aiohttp issue (https://github.com/aio-libs/aiohttp/issues/3122).

**Comment by cjavad** [comment]
I will get around to adding some test cases in a few weeks, but anyone is more than welcome to chip in, you can see my original standalone tests (especially the first one) in the issue.

**Comment by cjavad** [comment]
GitHub decided to downscale all their services to 0 so the previous test run "All succeeded" is marked as failed, but the test is valid and works.

**Review by 1st1 — APPROVED** [review]

**Comment by cjavad** [comment]
Is it realistic to expect this bug fix will be backported to 3.12? Or should we expect it in 3.13 or 3.14?

**Review by willingc — APPROVED** [review]
These changes look good to me.

@1st1 Double-checking with you that backporting to 3.13 and 3.12 makes sense for this PR.

**Comment by 1st1** [comment]
> @1st1 Double-checking with you that backporting to 3.13 and 3.12 makes sense for this PR.

I did a cursory review of this and read the sslproto.py code. I *think* this change is correct is should be ok to port to 3.13/12.

**Comment by willingc** [comment]
As a courtesy, I will keep this PR open a few more days before merging until @kumaraditya303 can take a look at this again.

File: Misc/NEWS.d/next/Core and Builtins/2024-05-12-03-10-36.gh-issue-118950.5Wc4vp.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Let _SSLProtocolTransport.is_closing reflect SSLProtocol internal transport
+closing state so that StreamWriter.drain will invoke sleep(0) which calls
+connection_lost and correctly notifies waiters of connection lost.
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Core and Builtins/2024-05-12-03-10-36.gh-issue-118950.5Wc4vp.rst:3]: Fix bug where SSLProtocol.connection_lost wasn't getting called when OSError was thrown on writing to socket.

Comment (kumaraditya303) [inline Misc/NEWS.d/next/Core and Builtins/2024-05-12-03-10-36.gh-issue-118950.5Wc4vp.rst]:
```suggestion
Fix bug where SSLProtocol.connection_lost wasn't getting called when OSError was thrown on writing to socket.
```

**Review by kumaraditya303 — APPROVED** [review]
LGTM, I tweaked the test a bit and the news entry.

**Comment by XenoAmess** [comment]
would it be backport to 3.11 & 3.10? many thanks...

**Comment by cjavad** [comment]
> would it be backport to 3.11 & 3.10? many thanks...
> 
> come to think of it, it might be somehow a security issue if you treat it as a way to perform dos

See https://github.com/SimplyPrint/printer-ws-client/blob/2f52cc85f604d42d5a6728f07cc029868102ae50/simplyprint_ws_client/_polyfill.py#L16 for an example of how to patch it dynamically.

**Comment by XenoAmess** [comment]
> > would it be backport to 3.11 & 3.10? many thanks...
> > come to think of it, it might be somehow a security issue if you treat it as a way to perform dos
> 
> See https://github.com/SimplyPrint/printer-ws-client/blob/2f52cc85f604d42d5a6728f07cc029868102ae50/simplyprint_ws_client/_polyfill.py#L16 for an example of how to patch it dynamically.

thank you sincerely

**Pull request merged**